### ISSUE TITLE
Using Ant test runner prevents console output in default ingredients

### ISSUE THREAD

**Issue by jtdaugherty** [ocharles/tasty-ant-xml#15]
If I run
```
main = do
  let ingredients = antXMLRunner : defaultIngredients
  defaultMainWithIngredients ingredients tests
```
then when tests fail, failures are recorded in the XML output (expected) but the console ingredient does not emit any output (unexpected).  Is this the intended behavior?  And, if so, can it be changed so that the console output is also emitted in the event of a test failure?

CC: @creswick @travitch

**Comment by ocharles** [comment]
I think this is the intended behavior, but I forget how ingredients interact. Pinging @feuerbach :)

**Comment by feuerbach** [comment]
By default, the first ingredient which "accepts" the options runs, and the rest are skipped.

It should be easy to write a function that combines multiple `TestReporter`s and returns a `TestReporter` that runs them together. Just take a look at the docs.

**Comment by ocharles** [comment]
Is that something `tasty-ant-xml` should be doing?

**Comment by feuerbach** [comment]
I believe the defaults are sensible. E.g. if you enable the ant-xml mode, then you are probably running this on a CI system and it displays the test results in its web UI, so there's not much point in printing it to stdout too.

Sometimes people like it differently, but that's why tasty is designed to be customizable by the end user. One doesn't need to patch tasty or tasty-ant-xml to get this feature; it can be done right in the test suite (or in a separate package, if it's so useful).

**Comment by creswick** [comment]
@feuerbach 
> It should be easy to write a function that combines multiple TestReporters and returns a TestReporter that runs them together. Just take a look at the docs.

A quick read over the docs makes it appear that doing this (running both via a function that combines TestReporters) would necessarily run the tests twice.  From the haddocks you linked to:

> There are two kinds of ingredients. TestReporter, if it agrees to run, automatically launches tests execution.

Is it possible to change that? We certainly don't want to run the tests twice just to get console output.

Would it be disruptive to separate text execution and output generation so that they can be handled separately? Is that something that the Ingredient system could support as-is, or would it need to be modified?

**Comment by feuerbach** [comment]
That's not quite right. When tasty runs a test reporter, it launches the tests, and passes the so-called status map to the reporter. The status map allows you to track the status of every test.
A simple way to think about test reporters as functions `StatusMap -> IO ()`. You have two such functions, `r1` and `r2`, and want to combine them. You do it like this:

`\sm -> do r1 sm; r2 sm`

This function *uses* `sm` twice, but the tests are run only once, which is what you want. Also, you may (or may not) choose to run them in parallel rather than sequentially.

This trick won't work with `TestManager`s, though, because those launch tests themselves by design.

> Would it be disruptive to separate text execution and output generation so that they can be handled separately?

And that's exactly how it works right now! The `StatusMap` is the bridge between these two independent processes.

**Comment by creswick** [comment]
> And that's exactly how it works right now! The StatusMap is the bridge between these two independent processes.

Ok, that's great! We'll give it a shot.

I would be nice to update the docs, and *fantastic* if there were a function provided that would combine Ingredients in a partial order, so that the API would support this kind of fan-out inherently. (Although I can't say I've thought about how you would sensibly combine TestManager's in that way.)

**Comment by feuerbach** [comment]
Can you point me to where exactly the docs are wrong or confusing, and how they should be improved?

**Comment by creswick** [comment]
@feuerbach 
The [haddocks](https://hackage.haskell.org/package/tasty-0.10.1.1/docs/Test-Tasty-Ingredients.html#t:Ingredient) you linked to for `Ingredient` state this (emphasis mine):

> There are two kinds of ingredients. TestReporter, if it agrees to run, **automatically launches tests execution.**
>
> TestManager is the second kind of ingredient. It is typically used for test management purposes (such as listing the test names), although it can also be used for running tests (but, **unlike TestReporter, it has to launch the tests manually**). It is therefore more general than TestReporter. TestReporter is provided just for convenience.
> 
> The **function's result should indicate whether all the tests passed**.

All the emphasized portions above heavily imply (to me) that `TestReporter`s actually run the tests, but that seems to contradict your comment above.

**Comment by feuerbach** [comment]
@creswick thanks. How about this? 

```
-- There are two kinds of ingredients.
--
-- The first kind is 'TestReporter'. If the ingredient that agrees to run
-- is a 'TestReporter', then tasty will automatically launch the tests and
-- pass a 'StatusMap' to the ingredient. All the ingredient needs to do
-- then is to process the test results and probably report them to the user
-- in some way (hence the name).
--
-- 'TestManager' is the second kind of ingredient. It is typically used for
-- test management purposes (such as listing the test names), although it
-- can also be used for running tests (but, unlike 'TestReporter', it has
-- to launch the tests manually if it wants them to be run).  It is
-- therefore more general than 'TestReporter'. 'TestReporter' is provided
-- just for convenience.
```

**Comment by creswick** [comment]
@feuerbach Looks good to me, thanks!

**Comment by ocharles** [comment]
It sounds like this is not a bug in `tasty-ant-xml` after all. @jtdaugherty if you disagree, feel free to reopen!

**Comment by jtdaugherty** [comment]
I agree.  I think the proposed documentation updates help, although even with those it isn't possible to understand what will happen in an ingredient list without observing the runtime results (i.e., the fact that you can put them in a list is misleading since the ones that come later can get effectively ignored).  But as you pointed out, that and the other issues are not about this library specifically.

Thanks!

**Comment by feuerbach** [comment]
@jtdaugherty: feel free to open further issues about the docs on tasty's issue tracker

**Comment by liskin** [comment]
The recent addition of [`composeReporters`](https://github.com/feuerbach/tasty/commit/e1f6a773ed42ad7401df2a8c3bbab1359dd0b22a) to tasty finally adds a decent way to enable both outputs easily.

Unfortunately when I stumbled upon this problem a few months ago, I didn't find this thread and neither did `composeReporters` exist and I wasn't smart enough to think of it myself. :-(
Instead I found some "ingredient transformer" stuff in `tasty-rerun` and when I got to implementing it, I ended up abusing that for this purpose, and [`tasty-jenkins-xml-0.1.0.0`](https://github.com/IxpertaSolutions/tasty-jenkins-xml/blob/170e55eac16d19ca240de16a3a90a18642703e63/src/Test/Tasty/Runners/JenkinsXML.hs) happened. @feuerbach corrected my course and I've rewritten my code to use `composeReporters`: https://github.com/IxpertaSolutions/tasty-jenkins-xml/blob/a368a800244343331f2809e3d5d988ee04c094f6/src/Test/Tasty/Runners/JenkinsXML.hs

There are still two additions (`--jxml` and `--exit-success`) that I think are useful in practice (well, they are in our env, may not be in any other) but it's small enough that I no longer feel it's worth a separate package. Do you think you may want to include some of that in `tasty-ant-xml`? I'm willing to submit pull requests, and I don't mind going the extra mile and adding tests and travis and hlint and README if you want any of that, as I've already done all that for `tasty-jenkins-xml`.
